File: Seori/2026/2026-06-W05/3699. Number of ZigZag Arrays I.py
class Solution:
    def zigZagArrays(self, n: int, l: int, r: int) -> int:
        MOD =  10 ** 9 + 7
        M = r - l

        # [1] dp 배열 정의. 
        # dp[v] = 현재 자리수에 값 v를 택할 때 만들 수 있는 M자 지그재그의 수
        # 계산의 편의를 위해 [l, r] -> [0, M(=r - l)]로 변환
        dp = [1] * (M + 1)
        
        # [2] 누적합을 통한 경우의 수 dp 배열 완성
        for i in range(1, n):
            next_dp = [0] * (M + 1)
            
            # M자 지그재그 -> 홀수 인덱스에서는 앞보다 큰 수
            if i % 2 == 1:
                # 앞에서부터 누적합 더하기
                running_sum = 0
                for v in range(M + 1):
                    next_dp[v] = running_sum
                    running_sum = (running_sum + dp[v]) % MOD

            # 짝수 인덱스에서는 앞보다 작은 수
            else:
                # 뒤에서부터 누적합 더하기
                running_sum = 0
                for v in range(M, -1, -1):
                    next_dp[v] = running_sum
                    running_sum = (running_sum + dp[v]) % MOD
                    
            dp = next_dp
            
        # [3] 모든 가능한 가짓수를 더하고, M자 W자 지그재그는 대칭관계이므로 *2하여 정답
        return sum(dp) * 2 % MOD

File: Lucy/src/2026/2026-06-W05/3699_Number_of_ZigZag_Arrays_I.ts
const MOD = 10 ** 9 + 7;

function zigZagArrays(n: number, l: number, r: number): number {
  const down = new Array<number>(r + 1).fill(0);
  const up = new Array<number>(r + 1).fill(0);
  const sumDown = new Array<number>(r + 2).fill(0);
  const sumUp = new Array<number>(r + 2).fill(0);

  // 값 초기화
  for (let i = l; i <= r; i++) {
    down[i] = up[i] = 1;
    sumDown[i] = sumUp[i] = i - l + 1; // sum[i] = l부터 i까지의 누적합, sum[i] = dp[l] + dp[l + 1] + ... + dp[i]
  }

  for (let i = 1; i < n; i++) {
    for (let j = l; j <= r; j++) {
      down[j] = (sumUp[r] - sumUp[j] + MOD) % MOD;
      up[j] = sumDown[j - 1];
    }

    sumDown[l] = down[l];
    sumUp[l] = up[l];

    for (let j = l + 1; j <= r; j++) {
      sumDown[j] = (sumDown[j - 1] + down[j]) % MOD;
      sumUp[j] = (sumUp[j - 1] + up[j]) % MOD;
    }
  }

  return (sumDown[r] + sumUp[r]) % MOD;
}
